### PR TITLE
Initial implementation of C# support

### PR DESCRIPTION
**Pull request by fabiocav**

### CODE REVIEW

**Comment by fabiocav** [comment]
@mathewc , @MikeStall opening this PR for review and feedback. This includes some changes from Mathew's initial commit.

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+                "Microsoft.WindowsAzure.Storage"
+            };
+
+        public static Assembly[] DefaultAssemblies
+            => new Assembly[]
+            {
+                typeof(TraceWriter).Assembly,
+                typeof(object).Assembly,
+                typeof(TimerInfo).Assembly
+            };
```
Comment (ahmelsayed) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: both `DefaultAssemblies` and `DefaultAssemblyReferences` seem like properties that can be cached for the lifetime of the host, right?

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+        public override async Task Invoke(object[] parameters)
+        {
+            object input = parameters[0];
+            TraceWriter traceWriter = (TraceWriter)parameters[1];
+            IBinder binder = (IBinder)parameters[2];
+            ExecutionContext functionExecutionContext = (ExecutionContext)parameters[3];
+
+            try
+            {
+                _fileTraceWriter.Verbose(string.Format("Function started"));
```
Comment (ahmelsayed) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: why `string.Format`?

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+                // parse the input as json to get the binding data
+                Dictionary<string, string> bindingData = new Dictionary<string, string>();
+                if (_outputBindings.Any(p => p.HasBindingParameters) ||
+                    _inputBindings.Any(p => p.HasBindingParameters))
+                {
+                    bindingData = GetBindingData(input);
+                }
+                bindingData["InvocationId"] = functionExecutionContext.InvocationId.ToString();
+
+                Dictionary<string, object> executionContext = new Dictionary<string, object>();
```
Comment (ahmelsayed) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: nit: `var`?

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs (hunk 1 above)
Comment (ahmelsayed) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: Also btw, what's the story for external dependencies?

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+                _fileTraceWriter.Verbose(string.Format("Function completed (Failure)"));
+                throw;
+            }
+        }
+
+        private MethodInfo GetFunctionTarget()
+        {
+            if (_function == null)
+            {
+                Script<object> script = CreateScript();
```
Comment (ahmelsayed) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: Does this mean we only support single file functions?

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs (hunk 1 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: Yes, those are temporary for now, but I'll change so the default ones are cached.

External references may be managed by using the "#r" directive, there's an example of that in a couple of scripts and you could do something like "#r "bin\MyAssembly.dll" to load an assembly deployed with the script. We'll also have a mechanism to install NuGet packages through a Project.json file. Packages referenced that way we'll have the assemblies automatically referenced when the script is compiled.

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs (hunk 2 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: I believe this was in the original code and I didn't bother changing. I'll make that change.

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs (hunk 4 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: This gets the function entry point. Additional files may be dynamically pulled in using the "#load "something.csx" directive.

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs (hunk 1 above)
Comment (ahmelsayed) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: so we're going to be using the project.json syntax. I was just worried that we would need to invent our own package restore and csproj like system to support external dependencies, class libraries, multi-file functions, etc. I played with roslyn auto compilation before and had to pass in an array of reference assemblies and an array of source files so it wasn't apparent how we would achieve that without inventing our own build conventions which I thought we should stay away from. but if there is a clear plan to use the project.json semantics then I'm fine with that.

**Comment by fabiocav** [comment]
@ahmedelnably just pushed a commit addressing some of the things you've pointed out. I also made a small change to BlobBinding and BindingContext to support what we need for C# bindings to work properly with all of the different types supported by different bindings. I've limited the change to that binding only, keeping the existing behavior for the other script invokers unchanged just so we can have an example of the kind of changes we'd need and we discuss how we want to move forward once @mathewc is available.

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs (hunk 1 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: Yes, we'll be using Project.json and the functionality provided to us by NuGet to install/restore packages and generate the lock file we'll be able to use to grab the references we'll need to pass into Roslyn (the last bit _may_ require some custom code, but we might be able to leverage the work other teams are doing, we're investigating), so we won't be reinventing the wheel with nugetsm, and external references work with the approach we're following as well.

**Comment by fabiocav** [comment]
Closing this as #25 has been opened against the csharp branch with the changes in this PR.

**Pull request closed without merging**